graphics/transparent_surface.cpp

```diff
@@ -606,49 +606,56 @@ TransparentSurface *TransparentSurface::rotoscale(const TransformStruct &transfo
 				dy = sh - dy;
 			}
 
-#ifdef ENABLE_BILINEAR
-			if ((dx > -1) && (dy > -1) && (dx < sw) && (dy < sh)) {
-				const tColorRGBA *sp = (const tColorRGBA *)getBasePtr(dx, dy);
-				tColorRGBA c00, c01, c10, c11, cswap;
-				c00 = *sp;
-				sp += 1;
-				c01 = *sp;
-				sp += (this->pitch / 4);
-				c11 = *sp;
-				sp -= 1;
-				c10 = *sp;
-				if (flipx) {
-					cswap = c00; c00=c01; c01=cswap;
-					cswap = c10; c10=c11; c11=cswap;
-				}
-				if (flipy) {
-					cswap = c00; c00=c10; c10=cswap;
-					cswap = c01; c01=c11; c11=cswap;
-				}
-				/*
-				* Interpolate colors
-				*/
-				int ex = (sdx & 0xffff);
-				int ey = (sdy & 0xffff);
-				int t1, t2;
-				t1 = ((((c01.r - c00.r) * ex) >> 16) + c00.r) & 0xff;
-				t2 = ((((c11.r - c10.r) * ex) >> 16) + c10.r) & 0xff;
-				pc->r = (((t2 - t1) * ey) >> 16) + t1;
-				t1 = ((((c01.g - c00.g) * ex) >> 16) + c00.g) & 0xff;
-				t2 = ((((c11.g - c10.g) * ex) >> 16) + c10.g) & 0xff;
-				pc->g = (((t2 - t1) * ey) >> 16) + t1;
-				t1 = ((((c01.b - c00.b) * ex) >> 16) + c00.b) & 0xff;
-				t2 = ((((c11.b - c10.b) * ex) >> 16) + c10.b) & 0xff;
-				pc->b = (((t2 - t1) * ey) >> 16) + t1;
-				t1 = ((((c01.a - c00.a) * ex) >> 16) + c00.a) & 0xff;
-				t2 = ((((c11.a - c10.a) * ex) >> 16) + c10.a) & 0xff;
-				pc->a = (((t2 - t1) * ey) >> 16) + t1;
-			}
-#else
+#ifdef DISABLE_BILINEAR
 			if ((dx >= 0) && (dy >= 0) && (dx < srcW) && (dy < srcH)) {
 				const tColorRGBA *sp = (const tColorRGBA *)getBasePtr(dx, dy);
 				*pc = *sp;
 			}
+#else
+			if (transform._filteringMode == FILTER_BILINEAR) {
```

Comment:
Performing this check inside the inner loop is slooooow, consider upping this to be a decision based on a template parameter instead, so that the compiler will create two versions of this function, thereby moving the branch outside the entire loop without having to duplicate the code. (I think we're already doing that in the various draw functions of TransparentSurface, otherwise you'll find examples in subr3v's TinyGL code from 2014).

Another alternative is to put the if outside the for-loop and have a new set of for-loops after, but that leads to code duplication.